<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Загрузим-необходимые-библиотеки" data-toc-modified-id="Загрузим-необходимые-библиотеки-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Загрузим необходимые библиотеки</a></span></li><li><span><a href="#Загрузим-данные" data-toc-modified-id="Загрузим-данные-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Загрузим данные</a></span></li><li><span><a href="#Изучим-данные-и-определим,-какая-обработка-потребуется" data-toc-modified-id="Изучим-данные-и-определим,-какая-обработка-потребуется-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Изучим данные и определим, какая обработка потребуется</a></span><ul class="toc-item"><li><span><a href="#Изучим-общую-информацию" data-toc-modified-id="Изучим-общую-информацию-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Изучим общую информацию</a></span></li><li><span><a href="#Изучим-сбалансированность-целевого-признака" data-toc-modified-id="Изучим-сбалансированность-целевого-признака-1.3.2"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>Изучим сбалансированность целевого признака</a></span></li></ul></li><li><span><a href="#Выводы-по-результатам-исследования-данных" data-toc-modified-id="Выводы-по-результатам-исследования-данных-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Выводы по результатам исследования данных</a></span></li><li><span><a href="#Проведем-подготовку-данных-для-моделирования" data-toc-modified-id="Проведем-подготовку-данных-для-моделирования-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Проведем подготовку данных для моделирования</a></span><ul class="toc-item"><li><span><a href="#Реиндексируем-массив-data" data-toc-modified-id="Реиндексируем-массив-data-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>Реиндексируем массив <code>data</code></a></span></li><li><span><a href="#Объявим-функцию-lemm_and_clear-для-очистки-и-лемматизации-выражений" data-toc-modified-id="Объявим-функцию-lemm_and_clear-для-очистки-и-лемматизации-выражений-1.5.2"><span class="toc-item-num">1.5.2&nbsp;&nbsp;</span>Объявим функцию <code>lemm_and_clear</code> для очистки и лемматизации выражений</a></span></li><li><span><a href="#Проверим-работу-функции:" data-toc-modified-id="Проверим-работу-функции:-1.5.3"><span class="toc-item-num">1.5.3&nbsp;&nbsp;</span>Проверим работу функции:</a></span></li><li><span><a href="#Лемматизируем-и-очистим-комментарии" data-toc-modified-id="Лемматизируем-и-очистим-комментарии-1.5.4"><span class="toc-item-num">1.5.4&nbsp;&nbsp;</span>Лемматизируем и очистим комментарии</a></span></li><li><span><a href="#Сохраним-лемматизированные-данные-на-диске-в-файле-lemm_data.csv" data-toc-modified-id="Сохраним-лемматизированные-данные-на-диске-в-файле-lemm_data.csv-1.5.5"><span class="toc-item-num">1.5.5&nbsp;&nbsp;</span>Сохраним лемматизированные данные на диске в файле <code>lemm_data.csv</code></a></span></li><li><span><a href="#Восстановим-данные-с-диска" data-toc-modified-id="Восстановим-данные-с-диска-1.5.6"><span class="toc-item-num">1.5.6&nbsp;&nbsp;</span>Восстановим данные с диска</a></span></li><li><span><a href="#Загрузим-перечень-не-значащих-слов-для-английского-языка" data-toc-modified-id="Загрузим-перечень-не-значащих-слов-для-английского-языка-1.5.7"><span class="toc-item-num">1.5.7&nbsp;&nbsp;</span>Загрузим перечень не значащих слов для английского языка</a></span></li><li><span><a href="#Создадим-из-data-корпус-из-лемматизированных-фраз-и-массив-целевого-признака" data-toc-modified-id="Создадим-из-data-корпус-из-лемматизированных-фраз-и-массив-целевого-признака-1.5.8"><span class="toc-item-num">1.5.8&nbsp;&nbsp;</span>Создадим из <code>data</code> корпус из лемматизированных фраз и массив целевого признака</a></span></li><li><span><a href="#Разделим-корпус-и-массив-целевого-признака-на-обучающую-и-тестовую-выборки" data-toc-modified-id="Разделим-корпус-и-массив-целевого-признака-на-обучающую-и-тестовую-выборки-1.5.9"><span class="toc-item-num">1.5.9&nbsp;&nbsp;</span>Разделим корпус и массив целевого признака на обучающую и тестовую выборки</a></span></li></ul></li><li><span><a href="#Заключение-по-результатам-обработки-исходных-данных-и-подготовке-признаков-для-моделирования" data-toc-modified-id="Заключение-по-результатам-обработки-исходных-данных-и-подготовке-признаков-для-моделирования-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Заключение по результатам обработки исходных данных и подготовке признаков для моделирования</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#Обучим-и-протестируем-модель-LogisticRegression-методом-кросс-валидации" data-toc-modified-id="Обучим-и-протестируем-модель-LogisticRegression-методом-кросс-валидации-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обучим и протестируем модель <code>LogisticRegression</code> методом <code>кросс-валидации</code></a></span><ul class="toc-item"><li><span><a href="#Зададим-неизменяемые-гиперпараметры-модели" data-toc-modified-id="Зададим-неизменяемые-гиперпараметры-модели-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Зададим неизменяемые гиперпараметры модели</a></span></li><li><span><a href="#Зададим-векторайзер" data-toc-modified-id="Зададим-векторайзер-2.1.2"><span class="toc-item-num">2.1.2&nbsp;&nbsp;</span>Зададим векторайзер</a></span></li><li><span><a href="#Объявим-шаги,-которые-будут-исполняться-в-Pipline" data-toc-modified-id="Объявим-шаги,-которые-будут-исполняться-в-Pipline-2.1.3"><span class="toc-item-num">2.1.3&nbsp;&nbsp;</span>Объявим шаги, которые будут исполняться в <code>Pipline</code></a></span></li><li><span><a href="#Зададим-диапазоны-гиперпараметров-для-поиска-оптимальной-метрики-f1-методом-кросс-валидации" data-toc-modified-id="Зададим-диапазоны-гиперпараметров-для-поиска-оптимальной-метрики-f1-методом-кросс-валидации-2.1.4"><span class="toc-item-num">2.1.4&nbsp;&nbsp;</span>Зададим диапазоны гиперпараметров для поиска оптимальной метрики <code>f1</code> методом <code>кросс-валидации</code></a></span></li><li><span><a href="#Объявим-функцию-поиска-оптимальных-гиперпараметров" data-toc-modified-id="Объявим-функцию-поиска-оптимальных-гиперпараметров-2.1.5"><span class="toc-item-num">2.1.5&nbsp;&nbsp;</span>Объявим функцию поиска оптимальных гиперпараметров</a></span></li><li><span><a href="#Проведем-поиск-оптимальных-гиперпараметров" data-toc-modified-id="Проведем-поиск-оптимальных-гиперпараметров-2.1.6"><span class="toc-item-num">2.1.6&nbsp;&nbsp;</span>Проведем поиск оптимальных гиперпараметров</a></span></li></ul></li><li><span><a href="#Обучим-и-протестируем-модель-SGDClassifier-методом-кросс-валидации" data-toc-modified-id="Обучим-и-протестируем-модель-SGDClassifier-методом-кросс-валидации-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обучим и протестируем модель <code>SGDClassifier</code> методом <code>кросс-валидации</code></a></span><ul class="toc-item"><li><span><a href="#Зададим-неизменяемые-гиперпараметры-модели" data-toc-modified-id="Зададим-неизменяемые-гиперпараметры-модели-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Зададим неизменяемые гиперпараметры модели</a></span></li><li><span><a href="#Объявим-шаги,-которые-будут-исполняться-в-Pipline" data-toc-modified-id="Объявим-шаги,-которые-будут-исполняться-в-Pipline-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Объявим шаги, которые будут исполняться в <code>Pipline</code></a></span></li><li><span><a href="#Зададим-диапазоны-гиперпараметров-для-поиска-оптимальнлй-метрики-f1-методом-кросс-валидации" data-toc-modified-id="Зададим-диапазоны-гиперпараметров-для-поиска-оптимальнлй-метрики-f1-методом-кросс-валидации-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>Зададим диапазоны гиперпараметров для поиска оптимальнлй метрики <code>f1</code> методом <code>кросс-валидации</code></a></span></li><li><span><a href="#Объявим-функцию-поиска-оптимальных-гиперпараметров" data-toc-modified-id="Объявим-функцию-поиска-оптимальных-гиперпараметров-2.2.4"><span class="toc-item-num">2.2.4&nbsp;&nbsp;</span>Объявим функцию поиска оптимальных гиперпараметров</a></span></li><li><span><a href="#Проведем-поиск-оптимальных-гиперпараметров" data-toc-modified-id="Проведем-поиск-оптимальных-гиперпараметров-2.2.5"><span class="toc-item-num">2.2.5&nbsp;&nbsp;</span>Проведем поиск оптимальных гиперпараметров</a></span></li></ul></li><li><span><a href="#Обучим,-подберем-гиперпараметры-и-протестируем-методом-кросс-валидации--модель-RandomForestClassifier" data-toc-modified-id="Обучим,-подберем-гиперпараметры-и-протестируем-методом-кросс-валидации--модель-RandomForestClassifier-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обучим, подберем гиперпараметры и протестируем методом <code>кросс-валидации</code>  модель <code>RandomForestClassifier</code></a></span><ul class="toc-item"><li><span><a href="#Зададим-неизменяемые-гиперпараметры-модели" data-toc-modified-id="Зададим-неизменяемые-гиперпараметры-модели-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Зададим неизменяемые гиперпараметры модели</a></span></li><li><span><a href="#Объявим-шаги,-которые-будут-исполняться-в-Pipline" data-toc-modified-id="Объявим-шаги,-которые-будут-исполняться-в-Pipline-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>Объявим шаги, которые будут исполняться в <code>Pipline</code></a></span></li><li><span><a href="#Зададим-диапазоны-гиперпараметров-для-поиска-оптимальнлй-метрики-f1-методом-кросс-валидации" data-toc-modified-id="Зададим-диапазоны-гиперпараметров-для-поиска-оптимальнлй-метрики-f1-методом-кросс-валидации-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span>Зададим диапазоны гиперпараметров для поиска оптимальнлй метрики <code>f1</code> методом <code>кросс-валидации</code></a></span></li><li><span><a href="#Объявим-функцию-поиска-оптимальных-гиперпараметров" data-toc-modified-id="Объявим-функцию-поиска-оптимальных-гиперпараметров-2.3.4"><span class="toc-item-num">2.3.4&nbsp;&nbsp;</span>Объявим функцию поиска оптимальных гиперпараметров</a></span></li><li><span><a href="#Проведем-поиск-оптимальных-гиперпараметров" data-toc-modified-id="Проведем-поиск-оптимальных-гиперпараметров-2.3.5"><span class="toc-item-num">2.3.5&nbsp;&nbsp;</span>Проведем поиск оптимальных гиперпараметров</a></span></li></ul></li><li><span><a href="#Обучим,-подберем-гиперпараметры-и-протестируем-методом-кросс-валидации-модель-LGBMClassifier" data-toc-modified-id="Обучим,-подберем-гиперпараметры-и-протестируем-методом-кросс-валидации-модель-LGBMClassifier-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Обучим, подберем гиперпараметры и протестируем методом <code>кросс-валидации</code> модель <code>LGBMClassifier</code></a></span><ul class="toc-item"><li><span><a href="#Зададим-неизменяемые-гиперпараметры-модели" data-toc-modified-id="Зададим-неизменяемые-гиперпараметры-модели-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>Зададим неизменяемые гиперпараметры модели</a></span></li><li><span><a href="#Зададим-векторайзер" data-toc-modified-id="Зададим-векторайзер-2.4.2"><span class="toc-item-num">2.4.2&nbsp;&nbsp;</span>Зададим векторайзер</a></span></li><li><span><a href="#Объявим-шаги,-которые-будут-исполняться-в-Pipline" data-toc-modified-id="Объявим-шаги,-которые-будут-исполняться-в-Pipline-2.4.3"><span class="toc-item-num">2.4.3&nbsp;&nbsp;</span>Объявим шаги, которые будут исполняться в <code>Pipline</code></a></span></li><li><span><a href="#Зададим-диапазоны-гиперпараметров-для-поиска-оптимальнлй-метрики-f1-методом-кросс-валидации" data-toc-modified-id="Зададим-диапазоны-гиперпараметров-для-поиска-оптимальнлй-метрики-f1-методом-кросс-валидации-2.4.4"><span class="toc-item-num">2.4.4&nbsp;&nbsp;</span>Зададим диапазоны гиперпараметров для поиска оптимальнлй метрики <code>f1</code> методом <code>кросс-валидации</code></a></span></li><li><span><a href="#Объявим-функцию-поиска-оптимальных-гиперпараметров" data-toc-modified-id="Объявим-функцию-поиска-оптимальных-гиперпараметров-2.4.5"><span class="toc-item-num">2.4.5&nbsp;&nbsp;</span>Объявим функцию поиска оптимальных гиперпараметров</a></span></li><li><span><a href="#Проведем-поиск-оптимальных-гиперпараметров" data-toc-modified-id="Проведем-поиск-оптимальных-гиперпараметров-2.4.6"><span class="toc-item-num">2.4.6&nbsp;&nbsp;</span>Проведем поиск оптимальных гиперпараметров</a></span></li></ul></li><li><span><a href="#Выводы-по-результатам-обучения-и-проверки-методом-кросс-валидации-моделей-LogisticClassifier,-SGDClassifier,-RandomForestClassifier-и-LGBMClasifier" data-toc-modified-id="Выводы-по-результатам-обучения-и-проверки-методом-кросс-валидации-моделей-LogisticClassifier,-SGDClassifier,-RandomForestClassifier-и-LGBMClasifier-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Выводы по результатам обучения и проверки методом <code>кросс-валидации</code> моделей <code>LogisticClassifier</code>, <code>SGDClassifier</code>, <code>RandomForestClassifier</code> и <code>LGBMClasifier</code></a></span></li><li><span><a href="#Тестирование-выбранной-модели" data-toc-modified-id="Тестирование-выбранной-модели-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Тестирование выбранной модели</a></span><ul class="toc-item"><li><span><a href="#Создадим-наборы-признаков-для-обучающей-и-тестовой-выборок" data-toc-modified-id="Создадим-наборы-признаков-для-обучающей-и-тестовой-выборок-2.6.1"><span class="toc-item-num">2.6.1&nbsp;&nbsp;</span>Создадим наборы признаков для <code>обучающей</code> и <code>тестовой</code> выборок</a></span><ul class="toc-item"><li><span><a href="#Зададим-найденные-при-обучении-моделей-оптимальные-гиперпараметры-параметры-для-модели,-создающей-векторы-из-корпуса" data-toc-modified-id="Зададим-найденные-при-обучении-моделей-оптимальные-гиперпараметры-параметры-для-модели,-создающей-векторы-из-корпуса-2.6.1.1"><span class="toc-item-num">2.6.1.1&nbsp;&nbsp;</span>Зададим найденные при обучении моделей оптимальные гиперпараметры параметры для модели, создающей векторы из корпуса</a></span></li><li><span><a href="#Обучим-модель-на-обучающем-корпусе-и-создадим-набор-признаков-в-обучающей-и-тестовой-выборках" data-toc-modified-id="Обучим-модель-на-обучающем-корпусе-и-создадим-набор-признаков-в-обучающей-и-тестовой-выборках-2.6.1.2"><span class="toc-item-num">2.6.1.2&nbsp;&nbsp;</span>Обучим модель на <code>обучающем</code> корпусе и создадим набор признаков в <code>обучающей</code> и <code>тестовой</code> выборках</a></span></li></ul></li><li><span><a href="#Тестирование-на-тестовой-выборке" data-toc-modified-id="Тестирование-на-тестовой-выборке-2.6.2"><span class="toc-item-num">2.6.2&nbsp;&nbsp;</span>Тестирование на тестовой выборке</a></span></li><li><span><a href="#Проверка-скорости-обучения-и-скорости-предсказаний-модели-LGBMClasifier-на-обучающей-выборке" data-toc-modified-id="Проверка-скорости-обучения-и-скорости-предсказаний-модели-LGBMClasifier-на-обучающей-выборке-2.6.3"><span class="toc-item-num">2.6.3&nbsp;&nbsp;</span>Проверка скорости обучения и скорости предсказаний модели <code>LGBMClasifier</code> на <strong>обучающей</strong> выборке</a></span></li><li><span><a href="#Проверка-модели-на-адекватность-путем-сравнения-метрики-f1-прогноза,-полученного-на-тестовой-выборке,-с-прогнозом,-состоящим-из-единиц-(все-комментарии-токсичные)" data-toc-modified-id="Проверка-модели-на-адекватность-путем-сравнения-метрики-f1-прогноза,-полученного-на-тестовой-выборке,-с-прогнозом,-состоящим-из-единиц-(все-комментарии-токсичные)-2.6.4"><span class="toc-item-num">2.6.4&nbsp;&nbsp;</span>Проверка модели на адекватность путем сравнения метрики <code>f1</code> прогноза, полученного на тестовой выборке, с прогнозом, состоящим из единиц (все комментарии токсичные)</a></span></li></ul></li><li><span><a href="#Выводы-по-результатам-тестирования-модели" data-toc-modified-id="Выводы-по-результатам-тестирования-модели-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Выводы по результатам тестирования модели</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Подготовка

### Загрузим необходимые библиотеки

In [15]:
# !pip install -U pip setuptools wheel
# !pip install -U spacy
# !spacy validate
# !spacy -m download en_core_web_trf

In [16]:
import pandas as pd
import numpy as np
import nltk
from nltk.corpus import stopwords as stopwords


from sklearn.feature_extraction.text import TfidfVectorizer

import re
import os
import spacy

from sklearn.linear_model import LogisticRegression, SGDClassifier
from sklearn.ensemble import RandomForestClassifier
from lightgbm import LGBMClassifier
from sklearn.dummy import DummyClassifier

from sklearn.pipeline import Pipeline

from sklearn.model_selection import (train_test_split,
                                     GridSearchCV)
from sklearn.metrics import classification_report, f1_score

from tqdm import tqdm

### Загрузим данные

In [17]:
path_1 = '/datasets/toxic_comments.csv'
path_2 = 'https://code.s3.yandex.net/datasets/toxic_comments.csv'
path_3 = '/home/beerbear/Загрузки/Lemmatization/toxic_comments.csv'


if os.path.exists(path_1):
    data = pd.read_csv(path_1, index_col=[0])
    print('Loaded:', path_1)

elif os.path.exists(path_2):
    data = pd.read_csv(path_2, index_col=[0])
    print('Loaded:', path_2)

elif os.path.exists(path_3):
    data = pd.read_csv(path_3, index_col=[0])
    print('Loaded:', path_3)
    

else:
    print('Файл не найден, проверьте путь')
    

Loaded: /home/beerbear/Загрузки/Lemmatization/toxic_comments.csv


### Изучим данные и определим, какая обработка потребуется

#### Изучим общую информацию

In [18]:
display(data.info())
display(data.head(3))
display(data.tail(3))
print('Дубликатов:', data.duplicated().sum())
display(data.describe())
print('Максимальное значение индекса:', data.index.max())
print('Количество уникальных индексов:', data.index.nunique())
print('Значения столбца toxic:', data['toxic'].unique())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 159292 entries, 0 to 159450
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159292 non-null  object
 1   toxic   159292 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 3.6+ MB


None

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0

text  toxic
159448  Spitzer \n\nUmm, theres no actual article for ...      0
159449  And it looks like it was actually you who put ...      0
159450  "\nAnd ... I really don't think you understand...      0

Дубликатов: 0


toxic
count  159292.000000
mean        0.101612
std         0.302139
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max         1.000000

Максимальное значение индекса: 159450
Количество уникальных индексов: 159292
Значения столбца toxic: [0 1]


#### Изучим сбалансированность целевого признака

In [19]:
print('Доля токсичных комментариев:',
    data['toxic'].mean())

Доля токсичных комментариев: 0.10161213369158527


### Выводы по результатам исследования данных

- в таблице **159292** объекта, один столбец с текстовыми выражениями и один столбец с **целевым** признаком `toxic`, который принимает значения `0` и `1`
- в индексе есть пропущенные значения, **реиндексируем** `data` для удобства дальнейшей обработки
- язык текстовых выражений - **английский**, имеются знаки препинания и символы, выражения не лемматизированы, требуется **лемматизация** и **очистка**
- дубликаты **отсутствуют**
- токсичные выражения составляют всего **10%** выборки, выборка **не сбалансирована**, что нужно будет учитывать при обучении

### Проведем подготовку данных для моделирования

#### Реиндексируем массив `data`

In [20]:
data = data.reset_index(drop = True)

In [21]:
data.tail(2)

text  toxic
159290  And it looks like it was actually you who put ...      0
159291  "\nAnd ... I really don't think you understand...      0

- проведено **реиндексирование** датасета `data`

#### Объявим функцию `lemm_and_clear` для очистки и лемматизации выражений

In [22]:
nlp = spacy.load("en_core_web_sm")

In [23]:
def lemm_and_clear(text):
    
    # лемматизируем текст
    
    doc = nlp(text)
    text = ' '.join([token.lemma_ for token in doc])

    
    # удалим небуквенные символы и лишние пробелы
     
    text = re.sub(r'[^a-zA-Z]', ' ', text)
    text = re.sub(r'   ', ' ', text)
    text = re.sub(r'   ', ' ', text)
    text = re.sub(r'  ', ' ', text)
    text = re.sub(r'  ', ' ', text)
    
   #  удалим пробелы в начале строки
   
    if text[0] == ' ':
        text = text[1:len(text)]
        
    return text

#### Проверим работу функции:

In [24]:
lemm_and_clear(' You should  ?  be ashamed of yourself went worked, \
               The striped bats are hanging on their feet for best ')

'you should be ashamed of yourself go work the stripe bat be hang on their foot for good'

In [25]:
data.loc[len(data) - 1, 'text']

'"\nAnd ... I really don\'t think you understand.  I came here and my idea was bad right away.  What kind of community goes ""you have bad ideas"" go away, instead of helping rewrite them.   "'

In [26]:
lemm_and_clear(data.loc[len(data) - 1,'text'])

'and I really do n t think you understand I come here and my idea be bad right away what kind of community go you have bad idea go away instead of helping rewrite they '

- проверка показала корректную **лемматизацию** и **очистку** комментариев

####  Лемматизируем и очистим комментарии

In [27]:
tqdm.pandas()

процесс занимает около 40 минут, поэтому файл создан один раз и записан на диск, потом читается с диска

In [28]:
data['lemm'] = data[
    'text'].str.lower().progress_apply(
    lemm_and_clear)

100%|███████████████████████████████████| 159292/159292 [37:27<00:00, 70.88it/s]


In [29]:
display(data.head(2), data.tail(2), data.shape)

text  toxic  \
0  Explanation\nWhy the edits made under my usern...      0   
1  D'aww! He matches this background colour I'm s...      0   

                                                lemm  
0  explanation why the edit make under my usernam...  
1  d aww he match this background colour I be see...

text  toxic  \
159290  And it looks like it was actually you who put ...      0   
159291  "\nAnd ... I really don't think you understand...      0   

                                                     lemm  
159290  and it look like it be actually you who put on...  
159291  and I really do n t think you understand I com...

(159292, 3)

- комментарии **лемматизированы** и **очищены**

#### Сохраним лемматизированные данные на диске в файле `lemm_data.csv`

In [30]:
data.to_csv('lemm_data.csv', index = True)

#### Восстановим данные с диска 

In [31]:
path_1 = 'lemm_data.csv'
path_2 = '/home/beerbear/lemm_data.csv'
path_3 = '/home/beerbear/Загрузки/Lemmatization/lemm_data.csv'


if os.path.exists(path_1):
    data = pd.read_csv(path_1, index_col=[0])
    print('Loaded:', path_1)

elif os.path.exists(path_2):
    data = pd.read_csv(path_2, index_col=[0])
    print('Loaded:', path_2)
    
elif os.path.exists(path_3):
    data = pd.read_csv(path_3, index_col=[0])
    print('Loaded:', path_3)

else:
    print('Файл не найден, проверьте путь')
    

Loaded: lemm_data.csv


In [32]:
display(data.head(2), data.tail(2), data.shape)

text  toxic  \
0  Explanation\nWhy the edits made under my usern...      0   
1  D'aww! He matches this background colour I'm s...      0   

                                                lemm  
0  explanation why the edit make under my usernam...  
1  d aww he match this background colour I be see...

text  toxic  \
159290  And it looks like it was actually you who put ...      0   
159291  "\nAnd ... I really don't think you understand...      0   

                                                     lemm  
159290  and it look like it be actually you who put on...  
159291  and I really do n t think you understand I com...

(159292, 3)

- создан и записан на диск датасет `lemm_data`, содержащий **лемматизированные и очищенные** от символов комментарии пользователей и целевой признак

#### Загрузим перечень не значащих слов для английского языка

In [33]:
nltk.download('stopwords')
stopwords = set(stopwords.words('english'))

[nltk_data] Downloading package stopwords to
[nltk_data]     /home/beerbear/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


#### Создадим из `data` корпус из лемматизированных фраз и массив целевого признака

In [34]:
corpus = data['lemm'].values.astype('U')
target = data['toxic']

In [35]:
display(corpus[0], corpus[len(corpus)-1], corpus.shape)
display(target.head(2), target.tail(2), target.shape)

'explanation why the edit make under my username hardcore metallica fan be revert they be n t vandalism just closure on some gas after I vote at new york dolls fac and please do n t remove the template from the talk page since I be retire now '

'and I really do n t think you understand I come here and my idea be bad right away what kind of community go you have bad idea go away instead of helping rewrite they '

(159292,)

0    0
1    0
Name: toxic, dtype: int64

159290    0
159291    0
Name: toxic, dtype: int64

(159292,)

In [36]:
del data

- из датафрейма `data` **созданы** корпус **лемматизированных и очищенных** комментариев и массив **целевого** признака

#### Разделим корпус и массив целевого признака на обучающую и тестовую выборки

In [37]:
corpus_train, corpus_test, \
target_train, target_test = train_test_split(corpus, target,
    random_state = 270923,
    shuffle = True,
    test_size = 0.25)

In [38]:
print(corpus_train.shape, round(target_train.shape[0]/corpus.shape[0], 2))
print(corpus_test.shape, round(target_test.shape[0]/corpus.shape[0], 2))
print(target_train.shape)
print(target_test.shape)

(119469,) 0.75
(39823,) 0.25
(119469,)
(39823,)


In [39]:
del corpus

### Заключение по результатам обработки исходных данных и подготовке признаков для моделирования

- созданы **корпуса лемматизированных и очищенных фраз** для создания обучающей и тестовой выборок признаков для обучения и тестирования моделей, а таже обучающая и тестовая выборки **целевого** признака
- размер **обучающей** выборки 75% от датасета, размер **тестовой** - 25%

## Обучение

### Обучим и протестируем модель `LogisticRegression` методом `кросс-валидации`

#### Зададим неизменяемые гиперпараметры модели

In [40]:
lr_model = LogisticRegression(
    solver = 'lbfgs', # 'lbfgs', на saga уходит в вечность, 1 шаг 74 минуты
    penalty= 'l2', #'elasticnet', для saga
    warm_start = True,
    random_state = 270923,
    class_weight = 'balanced',
    n_jobs = -1, 
    max_iter = 1000)

#### Зададим векторайзер

In [41]:
count_tf_idf = TfidfVectorizer(stop_words=stopwords)

#### Объявим шаги, которые будут исполняться в `Pipline`

In [42]:
lr_pipe = Pipeline(steps=[
   ("vectorizer", count_tf_idf),
   ("estimator", lr_model)])

#### Зададим диапазоны гиперпараметров для поиска оптимальной метрики `f1` методом `кросс-валидации`

In [43]:
lr_param_grid = {
    "vectorizer__min_df": [1], # 1, 2, 6],
    #"estimator__l1_ratio": [0.1, 0.15, 0.5],
    "estimator__C": [7.2] # 1 - 10
}

#### Объявим функцию поиска оптимальных гиперпараметров

In [44]:
lr_search = GridSearchCV(
    lr_pipe, 
    lr_param_grid,
    n_jobs= -1,
    scoring = 'f1',
    cv = 5,
    verbose = 10
)

#### Проведем поиск оптимальных гиперпараметров

In [45]:
lr_search.fit(corpus_train, target_train)

Fitting 5 folds for each of 1 candidates, totalling 5 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   2 out of   5 | elapsed:   52.9s remaining:  1.3min
[Parallel(n_jobs=-1)]: Done   3 out of   5 | elapsed:   54.2s remaining:   36.2s
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   58.1s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   58.1s finished


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('vectorizer',
                                        TfidfVectorizer(stop_words={'a',
                                                                    'about',
                                                                    'above',
                                                                    'after',
                                                                    'again',
                                                                    'against',
                                                                    'ain',
                                                                    'all', 'am',
                                                                    'an', 'and',
                                                                    'any',
                                                                    'are',
                                                                    'aren',
    

In [46]:
print("\nЛучшее полученное методом кросс-валидации \n \
значение метрики f1 = %0.4f:" % lr_search.best_score_,
"найдено при следующих параметрах")
print(lr_search.best_params_)


Лучшее полученное методом кросс-валидации 
 значение метрики f1 = 0.7645: найдено при следующих параметрах
{'estimator__C': 7.2, 'vectorizer__min_df': 1}


- полученное методом кросс-валидации значение метрики `f1` модели `LogisticRegression` составляет **0.7645**, модель **соответствует** требованиям задания (значение метрики `f1` должно быть не хуже 0.75)

### Обучим и протестируем модель `SGDClassifier` методом `кросс-валидации`

#### Зададим неизменяемые гиперпараметры модели

In [47]:
sgd_model = SGDClassifier(
    warm_start = True,
    penalty = 'elasticnet',
    random_state = 270923,
    class_weight = 'balanced',
    n_jobs = -1, 
    max_iter = 2000,
    early_stopping=True,)
    #tol = 0.00001)

#### Объявим шаги, которые будут исполняться в `Pipline`

In [48]:
sgd_pipe = Pipeline(steps=[
   ("vectorizer", count_tf_idf),
   ("estimator", sgd_model)])

#### Зададим диапазоны гиперпараметров для поиска оптимальнлй метрики `f1` методом `кросс-валидации`

In [49]:
sgd_param_grid = {
    "vectorizer__min_df": [3], # 2, 3, 6],
    "estimator__l1_ratio": [0.64], # 0.1 - 1.0
    #"estimator__C": [7.2] # 1 - 10
}

#### Объявим функцию поиска оптимальных гиперпараметров

In [50]:
sgd_search = GridSearchCV(
    sgd_pipe, 
    sgd_param_grid,
    n_jobs=-1,
    scoring = 'f1',
    cv = 5,
    verbose = 10
)

#### Проведем поиск оптимальных гиперпараметров

In [51]:
sgd_search.fit(corpus_train, target_train)

Fitting 5 folds for each of 1 candidates, totalling 5 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   2 out of   5 | elapsed:   15.5s remaining:   23.3s
[Parallel(n_jobs=-1)]: Done   3 out of   5 | elapsed:   16.4s remaining:   10.9s
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   17.3s remaining:    0.0s
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:   17.3s finished


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('vectorizer',
                                        TfidfVectorizer(stop_words={'a',
                                                                    'about',
                                                                    'above',
                                                                    'after',
                                                                    'again',
                                                                    'against',
                                                                    'ain',
                                                                    'all', 'am',
                                                                    'an', 'and',
                                                                    'any',
                                                                    'are',
                                                                    'aren',
    

In [52]:
print("\nЛучшее полученное методом кросс-валидации \n \
значение метрики f1 =%0.4f:" % sgd_search.best_score_,
"найдено при следующих параметрах")
print(sgd_search.best_params_)


Лучшее полученное методом кросс-валидации 
 значение метрики f1 =0.7448: найдено при следующих параметрах
{'estimator__l1_ratio': 0.64, 'vectorizer__min_df': 3}


- полученное методом кросс-валидации значение метрики `f1` модели `SGDClassifier` составляет **0.7448**, модель **не соответствует** требованиям задания (значение метрики `f1` должно быть не хуже 0.75),  модель `SGDClassifier` исключается из дальнейшего тестирования

### Обучим, подберем гиперпараметры и протестируем методом `кросс-валидации`  модель `RandomForestClassifier`

#### Зададим неизменяемые гиперпараметры модели

In [53]:
rfc_model = RandomForestClassifier(
    random_state = 270923,
    n_jobs = -1,
    class_weight = 'balanced')           

#### Объявим шаги, которые будут исполняться в `Pipline`

In [54]:
rfc_pipe = Pipeline(steps=[
   ("vectorizer", count_tf_idf),
   ("estimator", rfc_model)])

#### Зададим диапазоны гиперпараметров для поиска оптимальнлй метрики `f1` методом `кросс-валидации`

In [55]:
rfc_param_grid = {
        'vectorizer__min_df': [7], # 1 - 10
        'estimator__n_estimators': [200], # 150 [1 - 350], после 100 улучшение несущественное
        'estimator__max_depth': [400], # 400 [1 - 500],
        'estimator__min_samples_leaf': [1], #19   [1 - 30]
        'estimator__min_samples_split': [57] #57 [2 - 50]
    }

#### Объявим функцию поиска оптимальных гиперпараметров

In [56]:
rfc_search = GridSearchCV(
                 rfc_pipe, 
                 rfc_param_grid, 
                 cv = 5, 
                 n_jobs = -1,
                 scoring = 'f1',
                 verbose = 10)

#### Проведем поиск оптимальных гиперпараметров

In [57]:
rfc_search.fit(corpus_train, target_train)

Fitting 5 folds for each of 1 candidates, totalling 5 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.


[CV] estimator__C=7.2, vectorizer__min_df=1 ..........................
[CV]  estimator__C=7.2, vectorizer__min_df=1, score=0.752, total=  48.5s
[CV] estimator__l1_ratio=0.64, vectorizer__min_df=3 ..................
[CV]  estimator__l1_ratio=0.64, vectorizer__min_df=3, score=0.739, total=  15.1s
[CV] estimator__l1_ratio=0.64, vectorizer__min_df=3 ..................
[CV]  estimator__l1_ratio=0.64, vectorizer__min_df=3, score=0.753, total=  13.6s
[CV] estimator__l1_ratio=0.64, vectorizer__min_df=3 ..................
[CV]  estimator__l1_ratio=0.64, vectorizer__min_df=3, score=0.747, total=  14.5s


[Parallel(n_jobs=-1)]: Done   2 out of   5 | elapsed:  4.7min remaining:  7.1min
[Parallel(n_jobs=-1)]: Done   3 out of   5 | elapsed:  4.8min remaining:  3.2min
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:  4.8min remaining:    0.0s
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:  4.8min finished


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('vectorizer',
                                        TfidfVectorizer(stop_words={'a',
                                                                    'about',
                                                                    'above',
                                                                    'after',
                                                                    'again',
                                                                    'against',
                                                                    'ain',
                                                                    'all', 'am',
                                                                    'an', 'and',
                                                                    'any',
                                                                    'are',
                                                                    'aren',
    

In [58]:
print("\nЛучшее полученное методом кросс-валидации \n \
значение метрики f1 = %0.4f:" % rfc_search.best_score_,
"найдено при следующих параметрах")
print(rfc_search.best_params_)


Лучшее полученное методом кросс-валидации 
 значение метрики f1 = 0.7236: найдено при следующих параметрах
{'estimator__max_depth': 400, 'estimator__min_samples_leaf': 1, 'estimator__min_samples_split': 57, 'estimator__n_estimators': 200, 'vectorizer__min_df': 7}


- **наилучшее** полученное методом кросс-валидации значение метрики `f1` модели `RandomForestClassifier` составляет **0.7236**, что **не соответствует** требованиям задания, модель `RandomForestClassifier` исключается из дальнейшего тестирования

### Обучим, подберем гиперпараметры и протестируем методом `кросс-валидации` модель `LGBMClassifier`

#### Зададим неизменяемые гиперпараметры модели

In [59]:
lgb_model = LGBMClassifier(
    random_state = 270923,
    n_jobs = -1,
    verbose = -1,
    class_weight = 'balanced',
    force_row_wise=True)

#### Зададим векторайзер

In [60]:
count_tf_idf = TfidfVectorizer(stop_words=stopwords)

#### Объявим шаги, которые будут исполняться в `Pipline`

In [61]:
lgb_pipe = Pipeline(steps=[
   ("vectorizer", count_tf_idf),
   ("estimator", lgb_model)])

#### Зададим диапазоны гиперпараметров для поиска оптимальнлй метрики `f1` методом `кросс-валидации`

In [62]:
lgb_param_grid = {
        'vectorizer__min_df': [1], # 1, [1-10],
        'estimator__learning_rate': [0.5],   # 0.5 [0.01 - 0.9]
        'estimator__n_estimators': [150], # 110 [1 - 350],
        'estimator__max_depth': [30], # 30 [2 - 100],
        'estimator__num_leaves': [50], # 50 [2 - 75]
    }

#### Объявим функцию поиска оптимальных гиперпараметров

In [63]:
lgb_search = GridSearchCV(
                 lgb_pipe, 
                 lgb_param_grid, 
                 cv = 5, 
                 n_jobs = -1,
                 scoring = 'f1',
                 verbose = 10)

#### Проведем поиск оптимальных гиперпараметров

In [64]:
lgb_search.fit(corpus_train, target_train)

Fitting 5 folds for each of 1 candidates, totalling 5 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.


[CV] estimator__max_depth=400, estimator__min_samples_leaf=1, estimator__min_samples_split=57, estimator__n_estimators=200, vectorizer__min_df=7 
[CV]  estimator__max_depth=400, estimator__min_samples_leaf=1, estimator__min_samples_split=57, estimator__n_estimators=200, vectorizer__min_df=7, score=0.717, total= 4.7min
[CV] estimator__learning_rate=0.5, estimator__max_depth=30, estimator__n_estimators=150, estimator__num_leaves=50, vectorizer__min_df=1 
[CV]  estimator__learning_rate=0.5, estimator__max_depth=30, estimator__n_estimators=150, estimator__num_leaves=50, vectorizer__min_df=1, score=0.776, total=13.3min


[Parallel(n_jobs=-1)]: Done   2 out of   5 | elapsed: 14.3min remaining: 21.5min


[CV] estimator__C=7.2, vectorizer__min_df=1 ..........................
[CV]  estimator__C=7.2, vectorizer__min_df=1, score=0.768, total=  50.4s
[CV] estimator__l1_ratio=0.64, vectorizer__min_df=3 ..................
[CV]  estimator__l1_ratio=0.64, vectorizer__min_df=3, score=0.741, total=  12.9s
[CV] estimator__max_depth=400, estimator__min_samples_leaf=1, estimator__min_samples_split=57, estimator__n_estimators=200, vectorizer__min_df=7 
[CV]  estimator__max_depth=400, estimator__min_samples_leaf=1, estimator__min_samples_split=57, estimator__n_estimators=200, vectorizer__min_df=7, score=0.710, total= 4.8min
[CV] estimator__learning_rate=0.5, estimator__max_depth=30, estimator__n_estimators=150, estimator__num_leaves=50, vectorizer__min_df=1 
[CV]  estimator__learning_rate=0.5, estimator__max_depth=30, estimator__n_estimators=150, estimator__num_leaves=50, vectorizer__min_df=1, score=0.770, total=14.3min


[Parallel(n_jobs=-1)]: Done   3 out of   5 | elapsed: 14.3min remaining:  9.6min


[CV] estimator__C=7.2, vectorizer__min_df=1 ..........................
[CV]  estimator__C=7.2, vectorizer__min_df=1, score=0.762, total=  51.9s
[CV] estimator__max_depth=400, estimator__min_samples_leaf=1, estimator__min_samples_split=57, estimator__n_estimators=200, vectorizer__min_df=7 
[CV]  estimator__max_depth=400, estimator__min_samples_leaf=1, estimator__min_samples_split=57, estimator__n_estimators=200, vectorizer__min_df=7, score=0.737, total= 4.8min
[CV] estimator__learning_rate=0.5, estimator__max_depth=30, estimator__n_estimators=150, estimator__num_leaves=50, vectorizer__min_df=1 
[CV]  estimator__learning_rate=0.5, estimator__max_depth=30, estimator__n_estimators=150, estimator__num_leaves=50, vectorizer__min_df=1, score=0.755, total=14.7min


[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed: 14.7min remaining:    0.0s
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed: 14.7min finished


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('vectorizer',
                                        TfidfVectorizer(stop_words={'a',
                                                                    'about',
                                                                    'above',
                                                                    'after',
                                                                    'again',
                                                                    'against',
                                                                    'ain',
                                                                    'all', 'am',
                                                                    'an', 'and',
                                                                    'any',
                                                                    'are',
                                                                    'aren',
    

In [65]:
print("\nЛучшее полученное методом кросс-валидации \n \
значение метрики f1 = %0.4f:" % lgb_search.best_score_,
"найдено при следующих параметрах")
print(lgb_search.best_params_)


Лучшее полученное методом кросс-валидации 
 значение метрики f1 = 0.7670: найдено при следующих параметрах
{'estimator__learning_rate': 0.5, 'estimator__max_depth': 30, 'estimator__n_estimators': 150, 'estimator__num_leaves': 50, 'vectorizer__min_df': 1}


- полученное методом кросс-валидации значение метрики `f1` модели `LGBMClassifier` составляет **0.7670**, что **соответствует** требованиям задания и **превосходит** остальные модели

### Выводы по результатам обучения и проверки методом `кросс-валидации` моделей `LogisticClassifier`, `SGDClassifier`, `RandomForestClassifier` и `LGBMClasifier`

- по результатам обучения и проверки методом кросс-валидации моделей `LogisticClassifier`, `SGDClassifier`,  `RandomForestClassifier` и `LGBMClasifier` наилучший результат метрики: `f1` = **0.7670** продемонстрировала модель `LGBMClasifier`

- модель `LGBMClasifier` **рекомендуется для тестирования** с целью проверки её возможности использования для определения **тональности** комментариев

- **гиперпараметры** модели `LGBMClasifier`, при которых достигнуты указанные значения метрики `f1`:
    - `class_weight` ='balanced',
    - `learning_rate` = 0.5,
    - `max_depth` = 30,
    - `n_estimators` = 150,
    - `num_leaves` = 50

### Тестирование выбранной модели

#### Создадим наборы признаков для `обучающей` и `тестовой` выборок

##### Зададим найденные при обучении моделей оптимальные гиперпараметры параметры для модели, создающей векторы из корпуса

In [66]:
count_tf_idf = TfidfVectorizer(stop_words=stopwords, min_df = 1)

##### Обучим модель на `обучающем` корпусе и создадим набор признаков в `обучающей` и `тестовой` выборках

In [67]:
features_train = count_tf_idf.fit_transform(corpus_train)

features_test = count_tf_idf.transform(corpus_test)

In [68]:
print(features_train.shape, round(
    features_train.shape[0]/target.shape[0], 2))
print(features_test.shape, round(
    features_test.shape[0]/target.shape[0], 2))

(119469, 129130) 0.75
(39823, 129130) 0.25


- для проведения тестирования созданы **обучающая** и **тестовая** выборки признаков для обучения моделей в пропорциях 75% и 25% от исходного массива данных

#### Тестирование на тестовой выборке

In [69]:
lgbmc_model = LGBMClassifier(
    random_state = 270923,
    n_jobs = -1,
    verbose = -1,
    class_weight = 'balanced',
    force_row_wise=True,
    
    learning_rate=0.5,
    max_depth=30,
    n_estimators=150,
    num_leaves=50,
    )

In [70]:
lgbmc_model.fit(features_train, target_train)

LGBMClassifier(class_weight='balanced', force_row_wise=True, learning_rate=0.5,
               max_depth=30, n_estimators=150, n_jobs=-1, num_leaves=50,
               random_state=270923, verbose=-1)

In [71]:
lgbmc_predictions_test = lgbmc_model.predict(features_test)

In [72]:
lgbmc_f1_score = f1_score(target_test, lgbmc_predictions_test)
print("\nЗначение метрики f1 на тестовой выборке: %0.4f" % lgbmc_f1_score, '\n')


Значение метрики f1 на тестовой выборке: 0.7744 



In [73]:
print(classification_report(target_test, lgbmc_predictions_test))

              precision    recall  f1-score   support

           0       0.98      0.97      0.97     35729
           1       0.76      0.79      0.77      4094

    accuracy                           0.95     39823
   macro avg       0.87      0.88      0.87     39823
weighted avg       0.95      0.95      0.95     39823



- по результатам тестирования на `тестовой` выборке модель `LGBMClasifier` показала значение метрики `f1` = **0.7744**, что соответствует требованиям задания `f1` не менее 0.75
- по результатам тестирования на `тестовой` выборке модель `LGBMClasifier` показала следующие результаты полноты и правильности предсказаний:
    - модель **правильно** предсказала **79%** всех токсичных комментариев, при этом **правильность** предсказания (токсичный или нет) составила **76%**
    - модель **правильно** предсказала **97%** не токсичных комментариев с точностью определения **98%**

#### Проверка скорости обучения и скорости предсказаний модели `LGBMClasifier` на **обучающей** выборке

In [74]:
%%timeit
lgbmc_model.fit(features_train, target_train)

[CV] estimator__C=7.2, vectorizer__min_df=1 ..........................
[CV]  estimator__C=7.2, vectorizer__min_df=1, score=0.766, total=  51.8s
[CV] estimator__l1_ratio=0.64, vectorizer__min_df=3 ..................
[CV]  estimator__l1_ratio=0.64, vectorizer__min_df=3, score=0.745, total=  14.2s
[CV] estimator__max_depth=400, estimator__min_samples_leaf=1, estimator__min_samples_split=57, estimator__n_estimators=200, vectorizer__min_df=7 
[CV]  estimator__max_depth=400, estimator__min_samples_leaf=1, estimator__min_samples_split=57, estimator__n_estimators=200, vectorizer__min_df=7, score=0.727, total= 4.8min
[CV] estimator__learning_rate=0.5, estimator__max_depth=30, estimator__n_estimators=150, estimator__num_leaves=50, vectorizer__min_df=1 
[CV]  estimator__learning_rate=0.5, estimator__max_depth=30, estimator__n_estimators=150, estimator__num_leaves=50, vectorizer__min_df=1, score=0.768, total=14.3min
[CV] estimator__C=7.2, vectorizer__min_df=1 ..........................
[CV]  estim

In [75]:
%%timeit
lgbmc_predictions_train = lgbmc_model.predict(features_train)

2.99 s ± 19.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


- средняя скорость обучения модели `LGBMClasifier` на **обучающей** выборке составляет около 40 секунд
- средняя скорость предсказания модели `LGBMClasifier` на **обучающей** выборке составляет 3 секунды

- модель `LGBMClasifier` **может** быть использования для определения тональности комментариев в реальном времени

#### Проверка модели на адекватность путем сравнения метрики `f1` прогноза, полученного на тестовой выборке, с прогнозом, состоящим из единиц (все комментарии токсичные)

In [76]:
dummy_model = DummyClassifier(strategy = 'constant', constant = 1)

In [77]:
dummy_model.fit(features_train, target_train)
dummy_predict = dummy_model.predict(features_test)
f1_dummy = f1_score(target_test, dummy_predict)
print("Метрика f1 фиктивной модели: %0.2f" % f1_dummy)

Метрика f1 фиктивной модели: 0.19


- метрика `f1` модели `LGBMClasifier` на **тестовой** выборке составляет **0.77**, наилучшая метрика **фиктивной** модели составляет **0.19**, модель `LGBMClasifier` **адекватна**

### Выводы по результатам тестирования модели

- по результатам тестирования на `тестовой` выборке модель `LGBMClasifier` показала значение метрики `f1` = **0.77**, что соответствует требованиям задания `f1` не менее 0.75
- по результатам тестирования на `тестовой` выборке модель `LGBMClasifier` показала следующие результаты полноты и правильности предсказаний:
    - модель **правильно** предсказала **80%** всех токсичных комментариев, при этом **правильность** предсказания (токсичный или нет) составила **75%**
    - модель **правильно** предсказала **97%** не токсичных комментариев с точностью определения **98%**

- средняя скорость обучения модели `LGBMClasifier` на **обучающей** выборке составляет 1 минуту 40 секунд
- средняя скорость предсказания модели `LGBMClasifier` на **обучающей** выборке составляет 0.7 секунды

- модель `LGBMClasifier` **может** быть использования для определения тональности комментариев в реальном времени

- метрика `f1` модели `LGBMClasifier` на **тестовой** выборке составляет **0.77**, наилучшая метрика **фиктивной** модели составляет **0.19**, модель `LGBMClasifier` **адекватна**

## Выводы

- **Модель** `LGBMClasifier` со следующими **гиперпарметрами**:

    - `class_weight` = `balanced`,
    - `learning_rate` = 0.5,
    - `max_depth` = 30,
    - `min_child_samples` = 2,
    - `n_estimators` = 150,
    - `num_leaves` = 50
    
**рекомендуется** для оперативного прогнозирования тональности комметариев интернет-магазина «Викишоп» 